modules/basics/src/main/java/com/opengamma/strata/basics/date/MarketTenor.java

```diff
@@ -0,0 +1,271 @@
+ * <ul>
+ * <li>Trade date, the date that the trade is agreed
+ * <li>Spot date, the base for date calculations, typically 2 business days after the trade date, known as the spot lag
+ * <li>Start date, the date on which accrual starts, generally the spot date unless forward starting
+ * <li>End date, the date on which accrual ends
+ * </ul>
+ * <p>
+ * The period from start date to end date is represented by {@link Tenor}.
+ * By contrast, the market tenor describes both the spot lag and tenor.
+ * <p>
+ * In most cases, the market tenor is the same as the tenor, with the the spot lag implied and included by market convention.
+ * However, for very short trades there are special cases used in the market:
+ * <ul>
+ * <li>ON - Overnight, from today to tomorrow, spot lag of 0 and tenor of 1 day
+ * <li>TN - Tomorrow-Next, from tomorrow to the next day, spot lag of 1 and tenor of 1 day
+ * <li>SN - Spot-Next, from spot to the next day, market conventional spot lag and tenor of 1 day
+ * <li>SW - Spot-Week, from spot for one week, market conventional spot lag and tenor of 1 week
+ * </ul>
+ * <p>
+ * Note that if the market conventional spot lag is 1 day, TN and SN would resolve to the same dates.
+ * Note also that SN and SW exist for clarity, they might also be expressed as 1D and 1W with the spot implied.
+ * Other date combinations are possible in theory but tend not to exist in the market, for example a three day
+ * trade starting tomorrow, something which might require a code like T3D.
+ * 
+ * <h4>Usage</h4>
+ * {@code Tenor} implements {@code TemporalAmount} allowing it to be directly added to a date:
+ * <pre>
+ *  LocalDate later = baseDate.plus(tenor);
+ * </pre>
+ */
+public final class MarketTenor
+    implements Comparable<MarketTenor>, Serializable {
+
+  /**
+   * Serialization version.
+   */
+  private static final long serialVersionUID = 1;
+
+  /**
+   * A tenor code for Overnight, meaning from today to tomorrow.
+   */
+  public static final MarketTenor ON = new MarketTenor("ON", Tenor.ofDays(1), 0);
+  /**
+   * A tenor code for Tomorrow-Next, meaning from tomorrow to the next day.
+   */
+  public static final MarketTenor TN = new MarketTenor("TN", Tenor.ofDays(1), 1);
+  /**
+   * A tenor code for Spot-Next, meaning from the spot date to the next day.
+   * The spot date is usually two working days after today, but this varies by currency.
+   */
+  public static final MarketTenor SN = new MarketTenor("SN", Tenor.ofDays(1), 2);
+  /**
+   * A tenor code for Spot-Week, meaning one week starting from the spot date.
+   * The spot date is usually two working days after today, but this varies by currency.
+   */
+  public static final MarketTenor SW = new MarketTenor("SW", Tenor.ofWeeks(1), 2);
+
+  /**
+   * The code of the tenor.
+   */
+  private final String code;
+  /**
+   * The tenor.
+   */
+  private final Tenor tenor;
+  /**
+   * The order.
+   */
+  private final int order;
```

Comment:
The order concept feels a bit too 'magic'... I get that it's trying to solve for the problem of some market tenors having a known number of days whilst others use the default spot lag but it takes some getting your head around.

Could use a private SpotType enum with values ZERO_DAYS, ONE_DAY and MARKET_STANDARD? You'd need a bit more code in adjustSpotLag to map to the right number of days for the non standard one.

Alternatively you could keep the order concept but associate it with the private enum values (or just use ordinal()). Might give you the best of both worlds in terms of making it easier to read and keeping the code in the methods simple

Reply:
An enum feels over the top, but I think a variable rename and a constant make it clearer.